### PR TITLE
Added NotNullOrEmpty assertion #872

### LINKED ISSUE #872: Add NotNullOrEmpty assertion
There are times when I am creating assertions like so:

```powershell
$propNotSet = -not $Assert.NullOrEmpty($TargetObject, 'prop1').Result;
$anotherAssertion = $Assert.GreaterOrEqual($TargetObject, 'prop2', 3).Result;

$Assert.Create(
    ($propNotSet -and $anotherAssertion),
    $LocalizedData.Something,
    ...
);
```

But I would like to use `AllOf` with a collection of assertions. I cannot just pass in `NullOrEmpty` as the check needs to be inverted and I resort to using `Assert.Create` with my own AND bool condition as shown above. 

Ideally including a `NotNullOrEmpty` assertion would mean I could do:

```powershell
$Assert.AllOf(
    $Assert.NotNullOrEmpty($TargetObject, 'prop1'),
    $Assert.GreaterOrEqual($TargetObject, 'prop2', 3)
).Reason(
    $LocalizedData.Something,
    ...
);
```

Which makes things easier to work with and makes the assertion much more clearer in my opinion. 

I see we have a `NotNull` assertion. Would make sense to include a `NotNullOrEmpty` asserton as well. 

Let me know if this is something I can include or if I'm missing something with `AllOf`.

Additionally from the `AllOf` method signature:

```csharp
public AssertResult AllOf(params AssertResult[] results)
```

I can only get away with passing a collection of assertions, which makes me believe including `NotNullOrEmpty` is needed to for this to work with `AllOf`.

### PR DESCRIPTION
**Pull request by ArmaanMcleod**
## PR Summary

Fix #872

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Code changes**
  - [x] Have unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Microsoft/PSRule/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: src/PSRule/Runtime/AssertResult.cs [hunk 1]
```diff
@@ -124,7 +124,7 @@ public AssertResult ReasonIf(bool condition, string text, params object[] args)
         /// <returns>Returns an array of reasons. This will always return null when the Value is true.</returns>
         public string[] GetReason()
         {
-            if (!Result || _Reason == null || _Reason.Count == 0)
+            if (Result || _Reason == null || _Reason.Count == 0)
```
Comment (ArmaanMcleod) [inline src/PSRule/Runtime/AssertResult.cs:127]: Had to change this to return an empty array if the result is `true`. Let me know if this is incorrect or I've missed something.

**Comment by ArmaanMcleod** [comment]
I guess an issue with this assertion is that it could be interpreted as two ways:

1. It's not null and it's not empty
2. It's not null or it is empty

The intention would be to have it just the inversion of `NullOrEmpty`. 

Happy to not include this if it brings more confusion than benefit, I've just seen some cases where it could come in handy. 

An alternative could be to introduce `Empty` and `NotEmpty` assertions for collections. Then one could atleast use that in conjuncton with `Null` and `NotNull`.

**Review by BernieWhite — COMMENTED** [review]
@ArmaanMcleod The implementation is fine. However how do we differ from `$Assert.HasFieldValue` or `hasValue` expression.

https://github.com/microsoft/PSRule/blob/main/docs/concepts/PSRule/en-US/about_PSRule_Assert.md#hasfieldvalue

https://github.com/microsoft/PSRule/blob/main/docs/concepts/PSRule/en-US/about_PSRule_Expressions.md#hasvalue

File: src/PSRule/Runtime/AssertResult.cs (hunk 1 above)
Comment (BernieWhite) [reply src/PSRule/Runtime/AssertResult.cs:127]: @ArmaanMcleod Yes this looks like a bug. Thanks for fixing it.

**Comment by ArmaanMcleod** [comment]
> @ArmaanMcleod The implementation is fine. However how do we differ from `$Assert.HasFieldValue` or `hasValue` expression.
> 
> https://github.com/microsoft/PSRule/blob/main/docs/concepts/PSRule/en-US/about_PSRule_Assert.md#hasfieldvalue
> 
> https://github.com/microsoft/PSRule/blob/main/docs/concepts/PSRule/en-US/about_PSRule_Expressions.md#hasvalue
> 
> Maybe if there is a difference from these it may made sense up update the implementation. But they should be similar.
> 

@BernieWhite Yeah good point, I think there is really no difference between `HasFieldValue` vs `NotNullOrEmpty`. I guess I was trying to bring in something that is opposite to `NullOrEmpty`, and `NotNullOrEmpty` seemed like something I could remember to use easier. But yeah, don't want to bring in something that just does something that `HasFieldValue` already does 😄. I will just remember to use `HasFieldValue`. 

Can close this PR, was a good exercise to see how these assertions are working internally. 

> In terms of empty vs null. I think there is work to do there to provide options for both, but ideally both of these should be similar under expressions vs PowerShell.

I think this is worth looking into. My initial thought was bringing in `Empty`/`NotEmpty`, which also facilates more flexibility and does allow someone to use both `NotNull` and `NotEmpty` if they find that more intuitive for PowerShell.

**Pull request closed without merging**